sdk/eventhub/azure-eventhub/azure/eventhub/_consumer.py

```diff
@@ -197,32 +196,41 @@ def _open(self):
     def receive(self):
         # type: () -> None
         retried_times = 0
-        last_exception = None
         max_retries = (
             self._client._config.max_retries  # pylint:disable=protected-access
         )
 
-        while retried_times <= max_retries:
-            try:
-                if self._open():
-                    self._handler.do_work()  # type: ignore
-                return
-            except Exception as exception:  # pylint: disable=broad-except
-                if (
-                    isinstance(exception, uamqp.errors.LinkDetach)
-                    and exception.condition == uamqp.constants.ErrorCodes.LinkStolen  # pylint: disable=no-member
-                ):
-                    raise self._handle_exception(exception)
-                if not self.running:  # exit by close
-                    return
-                if self._last_received_event:
-                    self._offset = self._last_received_event.offset
-                last_exception = self._handle_exception(exception)
-                retried_times += 1
-                if retried_times > max_retries:
-                    _LOGGER.info(
-                        "%r operation has exhausted retry. Last exception: %r.",
-                        self._name,
-                        last_exception,
-                    )
-                    raise last_exception
+        if not self._message_buffer:  # then fetch some messages into buffer
+            while retried_times <= max_retries:
+                try:
+                    if self._open():
+                        self._handler.do_work()  # type: ignore
+                    break
+                except Exception as exception:  # pylint: disable=broad-except
+                    if (
+                        isinstance(exception, uamqp.errors.LinkDetach)
+                        and exception.condition == uamqp.constants.ErrorCodes.LinkStolen  # pylint: disable=no-member
+                    ):
+                        raise self._handle_exception(exception)
+                    if not self.running:  # exit by close
+                        return
+                    if self._last_received_event:
+                        self._offset = self._last_received_event.offset
+                    last_exception = self._handle_exception(exception)
+                    retried_times += 1
+                    if retried_times > max_retries:
+                        _LOGGER.info(
+                            "%r operation has exhausted retry. Last exception: %r.",
+                            self._name,
+                            last_exception,
+                        )
+                        raise last_exception
+
+        if self._message_buffer:
```

Comment:
I'm wondering if we should retrieve more than one here... don't we in the async version?

Reply:
The question is, if there are 8 partitions and prefetch is 300. By retrieving only one here, the EP will be working like, p0 1 event, p1 1 event, ... p7 1 event, then p0 1 event again ...

If we retrieve all events here, the EP will be working like p0 300 events, p1 300 events, ... p7 300 events, then p0 300 events again ...

The on_event callback is on single event, so the former will work in a more balanced way so I prefer the former. The latter is OK to me though. 
In the future on_event_batch, we can retrieve max_batch_size of events instead of all events. on_event is equivalent to batch size 1.

Reply:
https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py#L196

Reply:
hmm but it's true that this might end up being too "blocking" for sync function - so it would be good to perf test this...

Reply:
Changed back to retrieve all events instead of one by one. Let's discuss perf improvement / test separately.